heft-plugins/heft-lint-plugin/src/SarifFormatter.ts

```diff
@@ -0,0 +1,325 @@
+  startLine?: number;
+  startColumn?: number;
+  endLine?: number;
+  endColumn?: number;
+  snippet?: {
+    text: string;
+  };
+}
+
+export interface IStaticAnalysisRules {
+  id: string;
+  name?: string;
+  shortDescription?: {
+    text: string;
+  };
+  fullDescription?: {
+    text: string;
+  };
+  defaultConfiguration?: {
+    level: 'note' | 'warning' | 'error';
+  };
+  helpUri?: string;
+  properties?: {
+    category?: string;
+    precision?: 'very-high' | 'high' | 'medium' | 'low';
+    tags?: string[];
+    problem?: {
+      severity?: 'recommendation' | 'warning' | 'error';
+      securitySeverity?: number;
+    };
+  };
+}
+
+export interface ISarifFile {
+  location: {
+    uri: string;
+  };
+}
+
+export interface ISuppressedAnalysis {
+  kind: string;
+  justification: string;
+}
+
+export interface ISarifLocation {
+  physicalLocation: ISarifPhysicalLocation;
+}
+
+export interface ISarifPhysicalLocation {
+  artifactLocation: {
+    uri: string;
+    index?: number;
+  };
+  region?: IRegion;
+}
+
+export interface ISarifRule {
+  id: string;
+  helpUri?: string;
+  shortDescription?: {
+    text: string;
+  };
+  properties?: {
+    category?: string;
+  };
+}
+
+interface IMessage extends TEslint.Linter.LintMessage {
+  suppressions?: ISuppressedAnalysis[];
+}
+
+const INTERNAL_ERROR_ID: 'ESL0999' = 'ESL0999';
+const SARIF_VERSION: '2.1.0' = '2.1.0';
+const SARIF_INFORMATION_URI: 'http://json.schemastore.org/sarif-2.1.0-rtm.5' =
+  'http://json.schemastore.org/sarif-2.1.0-rtm.5';
+/**
+ * Converts ESLint results into a SARIF (Static Analysis Results Interchange Format) log.
+ *
+ * This function takes in a list of ESLint lint results, processes them to extract
+ * relevant information such as errors, warnings, and suppressed messages, and
+ * outputs a SARIF log which conforms to the SARIF v2.1.0 specification.
+ *
+ * @param results - An array of lint results from ESLint that contains linting information,
+ *                  such as file paths, messages, and suppression details.
+ * @param options - An object containing options for formatting:
+ *                  - `ignoreSuppressed`: Boolean flag to decide whether to ignore suppressed messages.
+ *                  - `eslintVersion`: Optional string to include the version of ESLint in the SARIF log.
+ * @returns The SARIF log containing information about the linting results in SARIF format.
+ */
+export function formatEslintResultsAsSARIF(
+  results: TEslint.ESLint.LintResult[],
+  rulesMeta: TEslint.ESLint.LintResultData['rulesMeta'],
+  options: ISerifFormatterOptions
+): ISarifLog {
+  const { ignoreSuppressed, eslintVersion, buildFolderPath } = options;
+  const toolConfigurationNotifications: ISarifRepresentation[] = [];
+  const sarifFiles: Map<string, ISarifFile> = new Map();
+  const sarifResults: ISarifRepresentation[] = [];
+  const sarifArtifactIndices: Map<ISarifFile, number> = new Map();
+  const sarifRules: Map<string, ISarifRule> = new Map();
```

Comment:
Similarly here, a simple array would be cleaner.

Reply:
Turned sarifRules from map to array and implemented similar functionality with the artifact indices.